R-package/tests/testthat/test-RGF_package.R

```diff
@@ -288,113 +283,257 @@ testthat::test_that("the 'mat_2scipy_sparse' returns a scipy CSC sparse matrix",
-  
+
     pr = init_regr$predict(scipySprse)
-  
+
     params = unlist(init_regr$get_params(deep = TRUE))
-  
+
     validate = names(params) %in% c("normalize", "loss", "verbose", "algorithm", "n_iter", "learning_rate",
                                     "sl2", "min_samples_leaf", "opt_interval", "l2", "n_tree_search",
                                     "reg_depth", "memory_policy", "test_interval", "max_leaf")
-  
+
     tmp_score = init_regr$score(x = scipySprse, y = y_reg)
-  
+
     tmp_score_W = init_regr$score(x = scipySprse, y = y_reg, sample_weight = W)
-  
+
     testthat::expect_true( length(pr) == length(y_reg) && sum(validate) == 15 && is.double(tmp_score) && is.double(tmp_score_W) )
   })
 }
+
+
+#=========================================================================================== test feature importances
+
+
+testthat::test_that("the feature importances of the 'RGF_Regressor' class works as expected", {
+  
+  skip_test_if_no_module("rgf.sklearn")
+  
+  init_regr = RGF_Regressor$new(max_leaf = 50, sl2 = 0.1, n_iter = 10)
+  
+  init_regr$fit(x = x_rgf, y = y_reg, sample_weight = W)                    # include also a vector of weights
+  
+  vec_imp = init_regr$feature_importances()
+
+  testthat::expect_true( inherits(vec_imp, 'array') && length(vec_imp) == ncol(x_rgf) )
+})
+
+
+#=========================================================================================== test dump-model
+
+
+testthat::test_that("the 'dump_model' method returns the correct output (Dumps the forest information to the R session -- works ONLY for RGF and NOT for FastRGF)", {
+  
+  skip_test_if_no_module("rgf.sklearn")
+  
+  init_class = RGF_Classifier$new(max_leaf = 50, sl2 = 0.1, n_iter = 10)
+  
+  init_class$fit(x = x_rgf, y = y_BINclass)
+  
+  dump_model = init_class$dump_model()
+  
+  #---------------------------------------
+  # for pretty-print in the R session use:  print( dump_model() )
+  #---------------------------------------
+  
+  output_dump = reticulate::py_capture_output(dump_model())
+  
+  testthat::expect_true( nchar(output_dump) > 0 && object.size(output_dump) > 0 )
+})
+
+
+#=========================================================================================== test saving a model
+
+
+testthat::test_that("the 'save_model' method returns the correct output -- works ONLY for RGF and NOT for FastRGF", {
+  
+  skip_test_if_no_module("rgf.sklearn")
+  
+  init_class = RGF_Classifier$new(max_leaf = 50, sl2 = 0.1, n_iter = 10)
+  
+  init_class$fit(x = x_rgf, y = y_BINclass)
+  
+  tmp_file = tempfile(fileext = '.model')
+  
+  SIZE_begin = file.info(tmp_file)$size
+  
+  sv_md = init_class$save_model(filename = tmp_file)
+  
+  SIZE_after = file.info(tmp_file)$size
+  
+  if (file.exists(tmp_file)) file.remove(tmp_file)
+  
+  testthat::expect_true( is.na(SIZE_begin) && (SIZE_after > 0) )
```

Comment:
@StrikerRUS, 

the only direct way to test that a model is valid would be to load the model (after saving it) by using a ```tryCatch()``` condition in R. If the loaded model gave an error then it would be invalid. 
However, I just noticed that the *rgf* function does have a ```save_model()``` but does not have a ```load_model()``` function. That means, I have two options,

- either to use the python *pickle* module from within R as is done in the [rgf python test](https://github.com/RGF-team/rgf/blob/master/python-package/tests/test_rgf_python.py#L39-L51) (that means I have to add an additional dependency and save the model as .pkl that means I won't utilize the save_model() function) or 

- to use ```from sklearn.exceptions import NotFittedError``` as is done in [another python tests](https://github.com/RGF-team/rgf/blob/master/python-package/tests/test_rgf_python.py#L349). Again I have to add the sklearn python dependency.

Please let me know what option do you think is the best one or if you have something else in your mind. thanks.

Reply:
I thinks it's too complicated way. Maybe simply read model file and do a text search for some keywords, like in https://github.com/RGF-team/rgf/pull/313#discussion_r353138413?

Reply:
@StrikerRUS, can you please elaborate on how to load the model (read the file) after I use the ```save_model()``` function, as there is no corresponding ```load_model()``` function. I can't think of something else other than use the ```pickle``` python module to save and load the model and utilize a ```tryCatch()``` function for error handling.
Let me know what you think. thanks.

Reply:
Oh, sorry - I forgot that model file is binary file actually. However, it seems that params are stored as raw string, so you can try to do the following:

```
import rgf
from sklearn.datasets import load_boston

X, y = load_boston(True)
est = rgf.RGFRegressor().fit(X, y)
est.save_model('test.model')

with open('test.model', 'rb') as fr:
    state = fr.read()

state

>>> b'\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00Q\x00\x00\x00M(\xc9\xfee\x886@\r\x00\x00\x00\xcd\x00\x00\x00NormalizeTarget,loss=LS,max_leaf_forest=500,test_interval=100,reg_L2=0.1,reg_sL2=0.1,reg_depth=1.0,min_pop=10,num_iteration_opt=10,num_tree_search=1,opt_interval=100,opt_stepsize=0.5,memory_policy=Generous\x00\n\x00\x00\x00-___-_RGF_\x00\x01\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x0043333s#@\x00\x00\x00\x00\x00\x00\x00\x00\x0c\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\xff\xff\xff\xff\xd4\xd4l{\x0f\x81\xce@\x00\x00\x00\x00\x00\x00\x00\x00\xdbp\xf1\x8c\x9a\xe9\xe1?\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x11\xbcB\tv\xad\xd9\xbf\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x00 ...

state.find(b'max_leaf_forest=') != -1
```

Reply:
Done!